File: clients/go/crypto/aeskw.go
package crypto

import (
	"crypto/aes"
	"encoding/binary"
	"errors"
)

// AES-256 Key Wrap (RFC 3394 / NIST SP 800-38F).
// This is a dev-only fallback for environments without the wolfcrypt shim.
//
// Strict/FIPS deployments MUST use the shim/HSM path (see node/provider_wolfcrypt.go).

var (
	kwDefaultIV = [8]byte{0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6}
)

// AESKeyWrapRFC3394 wraps plaintext key material using AES-KW.
// kek must be 32 bytes. keyIn must be 16..4096 bytes and a multiple of 8 bytes.
func AESKeyWrapRFC3394(kek, keyIn []byte) ([]byte, error) {
	if len(kek) != 32 {
		return nil, errors.New("aeskw: kek must be 32 bytes (AES-256)")
	}
	if len(keyIn) < 16 || len(keyIn) > 4096 || len(keyIn)%8 != 0 {
		return nil, errors.New("aeskw: keyIn must be 16..4096 bytes and multiple of 8")
	}

	block, err := aes.NewCipher(kek)
	if err != nil {
		return nil, err
	}

	n := len(keyIn) / 8
	r := make([][8]byte, n)
	for i := 0; i < n; i++ {
		copy(r[i][:], keyIn[i*8:(i+1)*8])
	}
	a := kwDefaultIV

	var b [16]byte
	for j := 0; j < 6; j++ {
		for i := 0; i < n; i++ {
			copy(b[0:8], a[:])
			copy(b[8:16], r[i][:])
			block.Encrypt(b[:], b[:])
			t := uint64(n*j + (i + 1))
			for k := 0; k < 8; k++ {
				a[k] = b[k] ^ byte(t>>(56-8*k))
			}
			copy(r[i][:], b[8:16])
		}
	}

	out := make([]byte, 0, 8+len(keyIn))
	out = append(out, a[:]...)
	for i := 0; i < n; i++ {
		out = append(out, r[i][:]...)
	}
	return out, nil
}

// AESKeyUnwrapRFC3394 unwraps AES-KW blob and returns plaintext key material.
// kek must be 32 bytes. wrapped must be 24..4104 bytes and multiple of 8 bytes.
func AESKeyUnwrapRFC3394(kek, wrapped []byte) ([]byte, error) {
	if len(kek) != 32 {
		return nil, errors.New("aeskw: kek must be 32 bytes (AES-256)")
	}
	if len(wrapped) < 24 || len(wrapped) > 4104 || len(wrapped)%8 != 0 {
		return nil, errors.New("aeskw: wrapped must be 24..4104 bytes and multiple of 8")
	}

	block, err := aes.NewCipher(kek)
	if err != nil {
		return nil, err
	}

	n := (len(wrapped) / 8) - 1
	var a [8]byte
	copy(a[:], wrapped[0:8])
	r := make([][8]byte, n)
	for i := 0; i < n; i++ {
		copy(r[i][:], wrapped[(i+1)*8:(i+2)*8])
	}

	var b [16]byte
	for j := 5; j >= 0; j-- {
		for i := n - 1; i >= 0; i-- {
			t := uint64(n*j + (i + 1))
			var aXor [8]byte
			copy(aXor[:], a[:])
			for k := 0; k < 8; k++ {
				aXor[k] ^= byte(t >> (56 - 8*k))
			}
			copy(b[0:8], aXor[:])
			copy(b[8:16], r[i][:])
			block.Decrypt(b[:], b[:])
			copy(a[:], b[0:8])
			copy(r[i][:], b[8:16])
		}
	}

	if a != kwDefaultIV {
		// RFC3394 integrity check.
		return nil, errors.New("aeskw: integrity check failed")
	}

	out := make([]byte, 0, n*8)
	for i := 0; i < n; i++ {
		out = append(out, r[i][:]...)
	}

	// Sanity: roundtrip check size
	if len(out)%8 != 0 {
		return nil, errors.New("aeskw: unwrap produced non-multiple-of-8 length")
	}
	// Keep a stable memory layout; no extra allocations from binary.
	_ = binary.BigEndian
	return out, nil
}

File: clients/rust/crates/rubin-node/src/aeskw.rs
// AES-256 Key Wrap (RFC 3394 / NIST SP 800-38F).
//
// This is a dev-only fallback for environments without a wolfcrypt shim keywrap
// implementation. Strict/FIPS deployments must use the shim/HSM path.

use aes::Aes256;
use aes::cipher::{BlockDecrypt, BlockEncrypt, KeyInit};

const DEFAULT_IV: [u8; 8] = [0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6];

fn xor_u64_be(a: &mut [u8; 8], t: u64) {
    for (k, v) in a.iter_mut().enumerate().take(8) {
        *v ^= ((t >> (56 - 8 * k)) & 0xff) as u8;
    }
}

pub fn aes_key_wrap_rfc3394(kek: &[u8], key_in: &[u8]) -> Result<Vec<u8>, String> {
    if kek.len() != 32 {
        return Err("aeskw: kek must be 32 bytes (AES-256)".into());
    }
    if key_in.len() < 16 || key_in.len() > 4096 || !key_in.len().is_multiple_of(8) {
        return Err("aeskw: keyIn must be 16..4096 bytes and multiple of 8".into());
    }

    let cipher = Aes256::new_from_slice(kek).map_err(|_| "aeskw: invalid kek".to_string())?;
    let n = key_in.len() / 8;

    let mut r = vec![[0u8; 8]; n];
    for i in 0..n {
        r[i].copy_from_slice(&key_in[i * 8..(i + 1) * 8]);
    }
    let mut a = DEFAULT_IV;

    for j in 0..6u64 {
        for (i, ri) in r.iter_mut().enumerate().take(n) {
            let mut b = [0u8; 16];
            b[0..8].copy_from_slice(&a);
            b[8..16].copy_from_slice(ri);

            let mut block = aes::cipher::generic_array::GenericArray::clone_from_slice(&b);
            cipher.encrypt_block(&mut block);
            b.copy_from_slice(&block);

            a.copy_from_slice(&b[0..8]);
            let t = (n as u64) * j + (i as u64) + 1;
            xor_u64_be(&mut a, t);
            ri.copy_from_slice(&b[8..16]);
        }
    }

    let mut out = Vec::with_capacity(8 + key_in.len());
    out.extend_from_slice(&a);
    for ri in r.iter().take(n) {
        out.extend_from_slice(ri);
    }
    Ok(out)
}

pub fn aes_key_unwrap_rfc3394(kek: &[u8], wrapped: &[u8]) -> Result<Vec<u8>, String> {
    if kek.len() != 32 {
        return Err("aeskw: kek must be 32 bytes (AES-256)".into());
    }
    if wrapped.len() < 24 || wrapped.len() > 4104 || !wrapped.len().is_multiple_of(8) {
        return Err("aeskw: wrapped must be 24..4104 bytes and multiple of 8".into());
    }

    let cipher = Aes256::new_from_slice(kek).map_err(|_| "aeskw: invalid kek".to_string())?;
    let n = (wrapped.len() / 8) - 1;

    let mut a = [0u8; 8];
    a.copy_from_slice(&wrapped[0..8]);
    let mut r = vec![[0u8; 8]; n];
    for i in 0..n {
        r[i].copy_from_slice(&wrapped[(i + 1) * 8..(i + 2) * 8]);
    }

    for j in (0..6u64).rev() {
        for i in (0..n).rev() {
            let t = (n as u64) * j + (i as u64) + 1;
            let mut a_xor = a;
            xor_u64_be(&mut a_xor, t);

            let mut b = [0u8; 16];
            b[0..8].copy_from_slice(&a_xor);
            b[8..16].copy_from_slice(&r[i]);

            let mut block = aes::cipher::generic_array::GenericArray::clone_from_slice(&b);
            cipher.decrypt_block(&mut block);
            b.copy_from_slice(&block);

            a.copy_from_slice(&b[0..8]);
            r[i].copy_from_slice(&b[8..16]);
        }
    }

    if a != DEFAULT_IV {
        return Err("aeskw: integrity check failed".into());
    }

    let mut out = Vec::with_capacity(n * 8);
    for ri in r.iter().take(n) {
        out.extend_from_slice(ri);
    }
    if out.len() % 8 != 0 {
        return Err("aeskw: unwrap produced non-multiple-of-8 length".into());
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_aeskw_roundtrip() {
        let kek = [0x11u8; 32];
        let key_in = [0x22u8; 16];
        let wrapped = aes_key_wrap_rfc3394(&kek, &key_in).unwrap();
        let unwrapped = aes_key_unwrap_rfc3394(&kek, &wrapped).unwrap();
        assert_eq!(unwrapped, key_in);
    }

    #[test]
    fn test_aeskw_wrong_kek_fails_integrity() {
        let kek = [0x11u8; 32];
        let kek2 = [0x12u8; 32];
        let key_in = [0x22u8; 16];
        let wrapped = aes_key_wrap_rfc3394(&kek, &key_in).unwrap();
        let err = aes_key_unwrap_rfc3394(&kek2, &wrapped).unwrap_err();
        assert!(err.contains("integrity"));
    }
}
